### ISSUE TITLE
Cluster does not initialize properly, kindnet times out and crashes.

### ISSUE THREAD

**Issue by salasrod** [kubernetes-sigs/kind#1461]
**What happened**:

CoreDNS fails to start, and Kindnet has timeouts in their logs:

```
I0402 17:39:58.069948       1 main.go:64] hostIP = 172.17.0.2
podIP = 172.17.0.2
I0402 17:40:33.166911       1 main.go:104] Failed to get nodes, retrying after error: Get https://10.96.0.1:443/api/v1/nodes: dial tcp 10.96.0.1:443: i/o timeout
I0402 17:41:03.167486       1 main.go:104] Failed to get nodes, retrying after error: Get https://10.96.0.1:443/api/v1/nodes: dial tcp 10.96.0.1:443: i/o timeout
I0402 17:41:34.167858       1 main.go:104] Failed to get nodes, retrying after error: Get https://10.96.0.1:443/api/v1/nodes: dial tcp 10.96.0.1:443: i/o timeout
I0402 17:42:06.168271       1 main.go:104] Failed to get nodes, retrying after error: Get https://10.96.0.1:443/api/v1/nodes: dial tcp 10.96.0.1:443: i/o timeout
I0402 17:42:39.168601       1 main.go:104] Failed to get nodes, retrying after error: Get https://10.96.0.1:443/api/v1/nodes: dial tcp 10.96.0.1:443: i/o timeout
panic: Reached maximum retries obtaining node list: Get https://10.96.0.1:443/api/v1/nodes: dial tcp 10.96.0.1:443: i/o timeout
```

**What you expected to happen**:

CoreDNS containers to start, and kindnet not to crash

**How to reproduce it (as minimally and precisely as possible)**:

`kind create cluster`
`kubectl -n kube-system logs kindnet-xxxxx`

```kubectl -n kube-system get pods -o wide
NAME                                         READY   STATUS             RESTARTS   AGE   IP           NODE                 NOMINATED NODE   READINESS GATES
coredns-6955765f44-tvtk5                     0/1     Pending            0          20m   <none>       <none>               <none>           <none>
coredns-6955765f44-vbpxc                     0/1     Pending            0          20m   <none>       <none>               <none>           <none>
etcd-kind-control-plane                      1/1     Running            0          20m   172.17.0.2   kind-control-plane   <none>           <none>
kindnet-wgsbc                                0/1     CrashLoopBackOff   5          20m   172.17.0.2   kind-control-plane   <none>           <none>
kube-apiserver-kind-control-plane            1/1     Running            0          20m   172.17.0.2   kind-control-plane   <none>           <none>
kube-controller-manager-kind-control-plane   1/1     Running            0          20m   172.17.0.2   kind-control-plane   <none>           <none>
kube-proxy-gwjfp                             1/1     Running            0          20m   172.17.0.2   kind-control-plane   <none>           <none>
kube-scheduler-kind-control-plane            1/1     Running            0          20m   172.17.0.2   kind-control-plane   <none>           <none>
```

**Environment:**

- kind version: (use `kind version`): 0.7.0
- Kubernetes version: (use `kubectl version`): 1.17.0
- Docker version: (use `docker info`): 19.03.8
- OS (e.g. from `/etc/os-release`): Gentoo

**Comment by BenTheElder** [comment]
kindnetd is failing here because it can't reach the apiserver, can you tell me more about your host environment and exactly how you ran kind?

**Comment by BenTheElder** [comment]
specifically `docker info` would be useful, and if you could dump `/etc/resolv.conf`, and `which iptables-save` / get the iptables version.

I suspect this is the upstream bug https://github.com/kubernetes/kubernetes/issues/71305, which has the following options:

- convert your host to use iptables-legacy instead of iptables-nft (see the bug)
- try using the latest kind from HEAD, where we do more to attempt to mitigate this

**Comment by salasrod** [comment]
`docker info`:

```
Client:
 Debug Mode: false
 Plugins:
  buildx: Build with BuildKit (Docker Inc.)

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 1
 Server Version: 19.03.8
 Storage Driver: overlay2
  Backing Filesystem: <unknown>
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 35bd7a5f69c13e1563af8a93431411cd9ecf5021
 runc version: 
 init version: fec3683b971d9c3ef73f284f176672c44b448662
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.5.14-gentoo
 Operating System: Gentoo/Linux
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 15.58GiB
 Name: thundercloud
 ID: DYOX:QA2L:J7ZA:WS2R:RRHG:5OCY:ZHGS:VBNH:HCJU:PZQJ:S4FC:HD65
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

`/etc/resolv.conf`
```
# Generated by NetworkManager
search lan
nameserver 192.168.1.20
nameserver 2604:5500:c248:6b00:134f:7b72:6bb3:71d2
nameserver 2604:5500:c248:6b00::1
```

I do have legacy iptables:

`eselect iptables list`
```
Available iptables symlink targets:
  [1]   xtables-legacy-multi *`
``` 
`iptables --version`
```
iptables v1.8.4 (legacy)
```

**Comment by salasrod** [comment]
Will try to use the kind from head and see what the results are and report back.

**Comment by aojea** [comment]
paste the `docker exec -it kind-control-plane iptables-save` please?

**Comment by salasrod** [comment]
`docker exec -it kind-control-plan iptables-save`

```
docker exec -it kind-control-plane iptables-save
# Generated by iptables-save v1.8.3 on Thu Apr  2 19:39:25 2020
*mangle
:PREROUTING ACCEPT [16594:3953361]
:INPUT ACCEPT [16594:3953361]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [16597:3953545]
:POSTROUTING ACCEPT [16597:3953545]
:KUBE-KUBELET-CANARY - [0:0]
COMMIT
# Completed on Thu Apr  2 19:39:25 2020
# Generated by iptables-save v1.8.3 on Thu Apr  2 19:39:25 2020
*filter
:INPUT ACCEPT [16594:3953361]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [16597:3953545]
:KUBE-FIREWALL - [0:0]
:KUBE-KUBELET-CANARY - [0:0]
COMMIT
# Completed on Thu Apr  2 19:39:25 2020
# Generated by iptables-save v1.8.3 on Thu Apr  2 19:39:25 2020
*nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:KUBE-KUBELET-CANARY - [0:0]
:KUBE-MARK-DROP - [0:0]
-A KUBE-MARK-DROP -j MARK --set-xmark 0x8000/0x8000
COMMIT
# Completed on Thu Apr  2 19:39:25 2020
```

**Comment by BenTheElder** [comment]
I think you might have to switch iptables to use the legacy backend. I'm not having this issue on a work machine that i'm pretty sure is nft backed though :thinking: 

I'll double check that config.

**Comment by BenTheElder** [comment]
google work machine (on which kind works...)
```
$ readlink -f $(which iptables)
/usr/sbin/xtables-nft-multi
```
the distro is customized debian-testing

**Comment by BenTheElder** [comment]
```
$ iptables-save --version
iptables-save v1.8.3 (nf_tables)
```

**Comment by BenTheElder** [comment]
So I'm pretty convinced that on iptables_nft systems the current mitigation is to switch to iptables legacy with update-alternatives and then reboot the host, which is not super great.

We can do better.
/assign
cc @howardjohn

**Comment by BenTheElder** [comment]
I've reproduced this on k8s HEAD after a report from John, and confirmed that the switch to legacy iptables mitigates.

**Comment by BenTheElder** [comment]
https://github.com/kubernetes-sigs/kind/pull/1508 should fix this.

**Comment by BenTheElder** [comment]
1508 is in, I've confirmed that it fixes it in my environment.

please re-open if you continue to see this.

**Comment by ntoofu** [comment]
I encountered the same problem. `kind` requires iptables to support `comment` match, so I have enabled `CONFIG_NETFILTER_XT_MATCH_COMMENT` in my kernel config to resolve the problem.

Most distributions probably enable it by default, but in the case of distributions like Gentoo Linux, you should configure it properly.

**Comment by BenTheElder** [comment]
I imagine that's kube-proxy in upstream Kubernetes requiring this. Kind inherets all upstream host requirements except swap disabled which we allow (though it causes issues).

**Comment by Tiberivs** [comment]
I encountered the same problem with running `kind` on Gentoo Linux. I discovered that `kind-proxy` requires
```
CONFIG_IP_VS_RR
CONFIG_IP_VS_WRR
CONFIG_NETFILTER_XT_MATCH_STATISTIC
CONFIG_NETFILTER_XT_MATCH_COMMENT
```

I think it is not exhaustive list of course, but may be helpful for someone.

**Comment by ensonic** [comment]
I've tested 0.11.1 and 0.12.0. Now kindnet and kube-proxy come up, but:
```
KUBECONFIG=/tmp/kubeconfig-618096523 kubectl logs local-path-provisioner-5ddd94ff66-nqwgb -n local-path-storage
time="2022-04-13T09:27:37Z" level=fatal msg="Error starting daemon: Cannot start Provisioner: failed to get Kubernetes server version: Get https://10.96.0.1:443/version?timeout=32s: x509: certificate signed by unknown authority"
```

**Comment by tin-pham** [comment]
> 

I know it've been 2 year but thanks man, i try to setup k8s and kind on gentoo, it come to the point I want to switch my os but you saved me

**Comment by tin-pham** [comment]
> I encountered the same problem with running `kind` on Gentoo Linux. I discovered that `kind-proxy` requires
> 
> ```
> CONFIG_IP_VS_RR
> CONFIG_IP_VS_WRR
> CONFIG_NETFILTER_XT_MATCH_STATISTIC
> CONFIG_NETFILTER_XT_MATCH_COMMENT
> ```
> 
> I think it is not exhaustive list of course, but may be helpful for someone.

I know it've been 2 year but thanks man, i try to setup k8s and kind on gentoo, it come to the point I want to switch my os but you saved me

**Comment by aitorpazos** [comment]
Same for me in Ubuntu 22.04, loading the following kernel modules did the trick:
```
sudo modprobe xt_comment xt_statistic ip_vs_rr ip_vs_wrr
```

**Comment by aojea** [comment]
> Same for me in Ubuntu 22.04, loading the following kernel modules did the trick:

ubuntu in WSL or the normal ubuntu distro?

**Comment by aitorpazos** [comment]
Actually KDE Neon on "bare metal" laptop

**Comment by chinchaun** [comment]
> I encountered the same problem with running `kind` on Gentoo Linux. I discovered that `kind-proxy` requires
> 
> ```
> CONFIG_IP_VS_RR
> CONFIG_IP_VS_WRR
> CONFIG_NETFILTER_XT_MATCH_STATISTIC
> CONFIG_NETFILTER_XT_MATCH_COMMENT
> ```
> 
> I think it is not exhaustive list of course, but may be helpful for someone.

Thank you very much! it solved my problem using a custom kernel 6.10
